metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala

```diff
@@ -1,44 +1,92 @@
 package scala.meta.internal.metals
 
 import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import java.util.Properties
 import scala.collection.mutable.ListBuffer
-import scala.util.control.NonFatal
+import scala.util.Try
 
 /**
  * Configuration that the user can override via workspace/didChangeConfiguration.
  *
  * @param javaHome The Java home location used to detect src.zip for JDK sources.
  */
 case class UserConfiguration(
-    javaHome: Option[String] = None
+    javaHome: Option[String] = None,
+    sbtLauncher: Option[String] = None,
+    sbtOpts: List[String] = Nil
 )
 
 object UserConfiguration {
+  def options: List[UserConfigurationOption] = List(
+    UserConfigurationOption(
+      "java-home",
+      "`JAVA_HOME` environment variable with fallback to `user.home` system property.",
+      "/Library/Java/JavaVirtualMachines/jdk1.8.0_192.jdk/Contents/Home",
+      "Java Home directory",
+      "The Java Home directory used for indexing JDK sources and locating the `java` binary."
+    ),
+    UserConfigurationOption(
+      "sbt-launcher",
+      "Metals embedded sbt-launch.jar.",
+      "/usr/local/Cellar/sbt/1.2.6/libexec/bin/sbt-launch.jar",
+      "sbt launcher jar",
+      "Optional sbt-launch.jar launcher to use when running `sbt bloopInstall`."
+    ),
+    UserConfigurationOption(
+      "sbt-options",
+      """empty string `""`.""",
+      "-Dsbt.override.build.repos=true -Divy.home=/home/ivy-cache",
+      "sbt JVM options",
+      """Additional space separated JVM options used for the `sbt bloopInstall` step.
+        |By default, Metals respects custom options in `.jvmopts` and `.sbtopts` of the workspace root directory,
+        |it's recommended to use those instead of customizing this setting. The benefit of `.jvmopts` and `.sbtopts`
+        |is that it's respected by other tools such as IntelliJ.
+        |""".stripMargin
+    )
+  )
+
   def fromJson(
-      base: UserConfiguration,
-      json: JsonObject
+      json: JsonObject,
+      properties: Properties = System.getProperties
   ): Either[List[String], UserConfiguration] = {
     val errors = ListBuffer.empty[String]
 
-    val javaHome: Option[String] =
-      if (json.has("javaHome")) {
-        try {
-          val string = json.get("javaHome").getAsString
-          if (string.isEmpty) None
-          else Some(string)
-        } catch {
-          case NonFatal(e) =>
-            errors += e.getMessage
-            base.javaHome
-        }
-      } else {
-        base.javaHome
+    def getKey(key: String): Option[String] = {
+      def option[T](fn: String => T): Option[T] =
+        Option(fn(key)).orElse(Option(fn(StringCase.kebabToCamel(key))))
+      for {
+        value <- option(json.get)
```

Comment:
Does this work?
The configuration sent by the client is nested under the `metals`, but here it's accessing `$key` instead of `metals.$key`. 

For example:

```json
{
  "jsonrpc": "2.0",
  "method": "workspace/didChangeConfiguration",
  "params": {
    "settings": {
      "metals": {
        "serverVersion": "0.4.0-SNAPSHOT",
        "serverProperties": "",
        "javaHome": ""
      }
    }
  }
}
```

Am I missing something?

Reply:
Good catch! This won't indeed work. I don't want to hard-code JSON decoding for the way VSCode sends it however. I think it's better to optionally drill into the `settings.metals` key but by default permit

```json
{
        "serverVersion": "0.4.0-SNAPSHOT",
        "serverProperties": "",
        "javaHome": ""
}
```

Reply:
Since that's not specified by LSP, why not advertising that nested structure?

Atom and Code seem to agree, do we have other data points?

Reply:
I've checked the protocol, and `settings` is mandated by it (I didn't know)

As per forcing the client in wrapping everything under `metals`, I don't see why not.

Emacs supports that LSP method and it allows the user to decide the structure, so lsp-scala can easily do it. I don't know yet about sublime and vim.

Habing